---
title: "Analysis"
author: "Rutuja"
date: "2024-02-27"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r, include=FALSE}
library(tidyverse)
library(lme4)
library(aod)
library(cowplot)
```

Reading the dataset

```{r}
df <- read.csv("final_data/d.csv")
```

Data info
'data.frame':	12071326 obs. of  36 variables:
 $ X             : int  1 2 3 4 5 6 7 8 9 10 ...
 $ chr_num       : int  1 1 1 1 1 1 1 1 1 1 ...
 $ pos           : num  1 2 3 4 5 6 7 8 9 10 ...
 $ ref           : chr  "C" "C" "A" "C" ...
 $ hap.wt        : num  0 0 0 0 0 0 0 0 0 0 ...
 $ hap.del       : num  0 0 0 0 0 0 0 0 0 0 ...
 $ dip.wt        : num  0 0 0 0 0 0 0 0 0 0 ...
 $ dip.del       : num  0 0 0 0 0 0 0 0 0 0 ...
 $ pwr.tot       : num  0 0 0 0 0 0 0 0 0 0 ...
 $ net_tot.minus : num  NA NA NA NA NA NA NA NA NA NA ...
 $ net_dens.minus: num  NA NA NA NA NA NA NA NA NA NA ...
 $ class.minus   : chr  NA NA NA NA ...
 $ net_tot.plus  : num  NA NA NA NA NA NA NA NA NA NA ...
 $ net_dens.plus : num  NA NA NA NA NA NA NA NA NA NA ...
 $ class.plus    : chr  NA NA NA NA ...
 $ pos500        : int  501 501 501 501 501 501 501 501 501 501 ...
 $ rep.ratio.hap : num  1.02 1.02 1.02 1.02 1.02 ...
 $ rep.slope.hap : num  -2.29e-05 -2.29e-05 -2.29e-05 -2.29e-05 -2.29e-05 ...
 $ rep.ratio.dip : num  1.04 1.04 1.04 1.04 1.04 ...
 $ rep.slope.dip : num  1.7e-05 1.7e-05 1.7e-05 1.7e-05 1.7e-05 ...
 $ rep.strand.hap: int  -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 ...
 $ rep.strand.dip: int  1 1 1 1 1 1 1 1 1 1 ...
 $ CHROM         : chr  NA NA NA NA ...
 $ REF           : chr  NA NA NA NA ...
 $ ALT           : chr  NA NA NA NA ...
 $ TYPE          : chr  NA NA NA NA ...
 $ RDH           : chr  NA NA NA NA ...
 $ MAT           : chr  NA NA NA NA ...
 $ ploidy        : chr  NA NA NA NA ...
 $ allele.freq   : num  NA NA NA NA NA NA NA NA NA NA ...
 $ chr_copy_num  : int  NA NA NA NA NA NA NA NA NA NA ...
 $ aneu11        : logi  NA NA NA NA NA NA ...
 $ indel.type    : chr  NA NA NA NA ...
 $ genic         : logi  NA NA NA NA NA NA ...
 $ mut.type      : chr  NA NA NA NA ...
 $ NCALLED       : int  NA NA NA NA NA NA NA NA NA NA ...
 
Building models, one column at a time.
  1. chr_num: identifier
  2. pos: position on the chromosome
  3. net_dens.minus: Average read density per base pair for the feature
  4. net_dens.plus: Same thing for plus strand
  5. NCALLED: number of mutations called at that locus
  
```{r}
# test1 <- df %>% select(chr_num, pos, net_dens.minus, net_dens.plus, NCALLED)
# 
# test1$net_dens.minus <- test1$net_dens.minus %>% replace_na(0) 
# test1$net_dens.plus <- test1$net_dens.plus %>% replace_na(0) 
# 
# test1$net_dens <- test1$net_dens.minus + test1$net_dens.plus
# test1 <- test1 %>% select(chr_num, pos, net_dens, NCALLED)
# test1$pos <- as.character(test1$pos)
# 
# test1 <- drop_na(test1)
```

Scaling before we can apply any model. I should not be scaling the label right?

```{r}
# test1.scaled <- test1
# test1.scaled$net_dens <- scale(test1$net_dens, center=TRUE, scale=TRUE)
# test1.scaled$NCALLED <- scale(test1$NCALLED, center=TRUE, scale=TRUE)
```

Using a linear model, linear mixed effects model

```{r}
# mod1.1 <- lm(NCALLED ~ chr_num, test1.scaled)
# summary(mod1.1)
# 
# mod1.2 <- lmer(NCALLED ~ net_dens + (1|chr_num), test1.scaled)
# summary(mod1.2)

```

No correlation detected. But correlation between chr_num and NCALLED

  1. chr_num: identifier
  2. pos: position on the chromosome
  3. net_dens.minus: Average read density per base pair for the feature
  4. net_dens.plus: Same thing for plus strand
  5. NCALLED: number of mutations called at that locus
  
```{r}
df$net_dens.minus <- df$net_dens.minus %>% replace_na(0) 
df$net_dens.plus <- df$net_dens.plus %>% replace_na(0) 

df$net_dens <- df$net_dens.minus + df$net_dens.plus
```


```{r}
# test2 <- df %>% select(chr_num, pos, net_dens, class.minus, class.plus, NCALLED) %>% filter(!is.na(NCALLED))
# 
# test2 %>% filter((!is.na(class.minus) & !is.na(class.plus) & !is.na(NCALLED)))
# 
# test2 <- test2 %>% mutate(class = case_when(
#   !is.na(class.minus) & is.na(class.plus) ~ class.minus,
#   is.na(class.minus) & !is.na(class.plus) ~ class.plus,
#   !is.na(class.minus) & !is.na(class.plus) ~ "both",
#   TRUE ~ "none"
# ))
# 
# test2 %>% filter((!is.na(class.minus) & !is.na(class.plus) & !is.na(NCALLED)))
# 
# test2 <- test2 %>% mutate(tr.or.not = case_when(
#   !is.na(class.minus) | !is.na(class.plus) ~ TRUE,
#   TRUE ~ FALSE
# ))
# 
# head(test2)
```

```{r}
# mod2.1 <- lm(NCALLED ~ class, test2)
# summary(mod2.1)
# 
# mod2.2 <- lmer(NCALLED ~ class + (1|chr_num), test2)
# summary(mod2.2)
```


```{r}
# mod2.3 <- lm(NCALLED ~ tr.or.not, test2)
# summary(mod2.3)
```

Now try this on the entire dataset

```{r}
df <- df %>% mutate(tr.or.not = case_when(
  !is.na(class.minus) | !is.na(class.plus) ~ TRUE,
  TRUE ~ FALSE
))

df <- df %>% mutate(class = case_when(
  !is.na(class.minus) & is.na(class.plus) ~ class.minus,
  is.na(class.minus) & !is.na(class.plus) ~ class.plus,
  !is.na(class.minus) & !is.na(class.plus) ~ "both",
  TRUE ~ "none"
))

df <- df %>% mutate(transcribed = ifelse(df$net_dens > 0, TRUE, FALSE))

df$net_tot <- df$net_tot.minus + df$net_dens.plus

```

There are a lot of NAs in d. Getting rid of most of them to make modelling easier

```{r}
df <- df %>% select(-CHROM)
df$net_tot.minus <- replace_na(df$net_tot.minus, 0)
df$net_tot.plus <- replace_na(df$net_tot.plus, 0)
df$class.minus <- replace_na(df$class.minus, "none")
df$class.plus <- replace_na(df$class.plus, "none")
df$net_tot.minus <- replace_na(df$net_tot.minus, 0)
df$net_tot.plus <- replace_na(df$net_tot.plus, 0)
df$allele.freq <- replace_na(df$allele.freq, 0)
df$genic <- replace_na(df$genic, FALSE)
df$NCALLED <- replace_na(df$NCALLED, 0)
df <- df %>% mutate(mutated = ifelse(df$NCALLED > 0, 1, 0))
```

NA and non-NA columns

```{r}
ready.cols <- sapply(df, function(x)all(!is.na(x)))
ready.cols[ready.cols]
```

Trying out models with the entire dataset

```{r}
# mod3.1 <- lm(NCALLED ~ net_dens, df)
# summary(mod3.1)
# 
# mod3.2 <- lm(NCALLED ~ genic, df)
# summary(mod3.2)
# 
# mod3.3 <- lm(NCALLED ~ transcribed, df)
# summary(mod3.3)
# 
# mod3.4 <- lm(NCALLED ~ net_tot, df)
# summary(mod3.4)
```

Type conversions and scaling

```{r}
# test.df <- df %>% select(chr_num, pos, allele.freq, genic, NCALLED, net_dens, tr.or.not, class, transcribed, net_tot)

# test.df <- df[, c("chr_num", "pos", "NCALLED", "mutated", "allele.freq", "genic", "class")]
# test.df$chr_num <- as.character(df$chr_num)
# test.df$net_dens <- scale(df$net_dens)
# # test.df$allele.freq <- scale(df$allele.freq)
# # test.df$genic <- scale(df$genic)
# test.df$net_tot <- scale(df$net_tot)
# test.df$tr.or.not <- scale(df$tr.or.not)
# test.df$transcribed <- scale(df$transcribed)

```


Same models but experimenting

```{r}
# mod4.1 <- lmer(NCALLED ~ 1 + net_dens + (1|chr_num), test.df)
# summary(mod4.1)
# 
# mod4.2 <- lm(NCALLED ~ net_dens, test.df)
# summary(mod4.2)
# 
# mod4.3 <- lm(NCALLED ~ net_tot, test.df)
# summary(mod4.3)
# 
# mod4.4 <- lm(NCALLED ~ tr.or.not, test.df)
# summary(mod4.4)
# 
# mod4.5 <- lm(NCALLED ~ transcribed, test.df)
# summary(mod4.5)
```

```{r}
# mod4.1 <- lmer(mutated ~ 1 + net_dens + (1|chr_num), test.df)
# summary(mod4.1)
# 
# mod4.2 <- lm(mutated ~ net_dens, test.df)
# summary(mod4.2)
# 
# mod4.3 <- lm(mutated ~ net_tot, test.df)
# summary(mod4.3)
# 
# mod4.4 <- lm(mutated ~ tr.or.not, test.df)
# summary(mod4.4)
# 
# mod4.5 <- lm(mutated ~ transcribed, test.df)
# summary(mod4.5)
```


```{r}
# mod4.2 <- glm(mutated ~ net_dens, family = "binomial", data = test.df)
# summary(mod4.2)
# 
# mod4.3 <- glm(mutated ~ net_tot, family = "binomial", data = test.df)
# summary(mod4.3)
# 
# mod4.4 <- glm(mutated ~ tr.or.not, family = "binomial", data = test.df)
# summary(mod4.4)
# 
# mod4.5 <- glm(mutated ~ transcribed, family = "binomial", data = test.df)
# summary(mod4.5)
```

Adding in multiple factors

```{r}
# mod <- glm(mutated ~ class + net_tot, family = "binomial", data=test.df)
# summary(mod)
```

Thinking about collisions

```{r}
collisions <- df %>% select(chr_num, pos, rep.strand.hap, rep.strand.dip, net_tot.minus, net_tot.plus, mutated)
```

So for the transcription direction, I am using the direction of whatever is greater between net_tot.plus and net_tot.minus. Finding all the collisions spots.

```{r}
collisions <- collisions %>% mutate(tr.strand = case_when(
  net_tot.minus > net_tot.plus ~ -1,
  net_tot.plus > net_tot.minus ~ 1
))
```

Replication profile plot

```{r}
df %>% filter(chr_num == 1) %>% ggplot() +
  geom_line(aes(x=pos, y=rep.ratio.hap), color='red') +
  geom_line(aes(x=pos, y=rep.ratio.dip), color='blue')

df %>% filter(chr_num == 1) %>% ggplot() +
  geom_line(aes(x=pos, y=rep.slope.hap), color='red') +
  geom_line(aes(x=pos, y=rep.slope.dip), color='blue')
```


Headon collisions: sign of replication and transcription is same
Cotranscriptional collisions: sign of replication and transcription is diff

```{r}
collisions <- collisions %>% mutate(collision.hap = case_when(
  rep.strand.hap * tr.strand == 1 ~ 1,
  TRUE ~ 0
))

collisions <- collisions %>% mutate(collision.dip = case_when(
  rep.strand.dip * tr.strand == 1 ~ 1,
  TRUE ~ 0
))

collisions <- collisions %>% mutate(net_tot = case_when(
  tr.strand == 1 ~ net_tot.plus,
  tr.strand == -1 ~ net_tot.minus,
  TRUE ~ 0
))

collisions <- collisions %>% mutate(net_tot.scaled = scale(net_tot))
```


```{r}
# Not significant

mod <- glm(mutated ~ collision.dip + net_tot, family = "binomial", collisions)
summary(mod)
```

```{r}
mod <- glm(mutated ~ collision.dip + net_tot.scaled, family = "binomial", collisions)
summary(mod)
```


```{r}
collisions$class <- df$class

mod <- glm(mutated ~ collision.dip + net_tot.scaled + class, family = "binomial", collisions)
summary(mod)

# Running a wald test too!

wald.test(b=coef(mod), Sigma = vcov(mod), Terms=4:20)
```


Testing out just class

```{r}
mod <- glm(mutated ~ class, family = "binomial", collisions)
summary(mod)

# Running a wald test too!

wald.test(b=coef(mod), Sigma = vcov(mod), Terms=2:18)
```

Testing out just net_tot.scaled again

```{r}
mod <- glm(mutated ~ net_tot.scaled, family = "binomial", collisions)
summary(mod)
```



Redoing those replication calculations

```{r}
# replication dynamics
rep.dynamics.hap<-read.delim("replication_data/Muller_rep_times_hap.txt")
rep.dynamics.hap$chr<-sapply(rep.dynamics.hap$chr_num,function(x){if(x<10){paste("chr0",x,sep="")}else{paste("chr",x,sep="")}})
rep.dynamics.hap$start<-rep.dynamics.hap$pos-500; rep.dynamics.hap$end<-rep.dynamics.hap$pos+499

rep.dynamics.dip<-read.delim("replication_data/Muller_rep_times_dip.txt")
rep.dynamics.dip$chr<-sapply(rep.dynamics.dip$chr_num,function(x){if(x<10){paste("chr0",x,sep="")}else{paste("chr",x,sep="")}})
rep.dynamics.dip$start<-rep.dynamics.dip$pos-500; rep.dynamics.dip$end<-rep.dynamics.dip$pos+499


rep.dynamics <- full_join(rep.dynamics.dip[,c("chr_num", "pos", "chr", "ratio")], rep.dynamics.hap[,c("chr_num", "pos", "chr", "ratio")], by=c("chr_num", "pos", "chr"), suffix = c(".dip", ".hap"))
rep.dynamics$ratio <- rowMeans(cbind(rep.dynamics$ratio.dip, rep.dynamics$ratio.hap), na.rm=TRUE)

```


```{r}
loess.span<-50000
# loess.models.hap<-list()
# for(chr_num in c(1:16)){
# 	chr.data<-rep.dynamics.hap[rep.dynamics.hap$chr_num==chr_num,]
# 	mod<-loess(ratio~pos, chr.data,span=loess.span/max(chr.data$pos))
# 	loess.models.hap<-c(loess.models.hap, list(mod))
# }
# loess.models.dip<-list()
# for(chr_num in c(1:16)){
# 	chr.data<-rep.dynamics.dip[rep.dynamics.dip$chr_num==chr_num,]
# 	mod<-loess(ratio~pos, chr.data,span=loess.span/max(chr.data$pos))
# 	loess.models.dip<-c(loess.models.dip, list(mod))
# }

loess.models <- list()
for(chr_num in c(1:16)){
	chr.data<-rep.dynamics[rep.dynamics$chr_num==chr_num,]
	mod<-loess(ratio~pos, chr.data,span=loess.span/max(chr.data$pos))
	loess.models<-c(loess.models, list(mod))
}
```


```{r}
chr_names2<-c("chr01","chr02","chr03","chr04","chr05","chr06","chr07","chr08","chr09","chr10","chr11","chr12","chr13","chr14","chr15","chr16")

nderiv<-function(fit, x, eps=1e-5){(predict(fit, x + eps) - predict(fit, x - eps))/(2 * eps)}

# nderiv will return NA at the boundary, so add or subtract 1 to get the slope
find.rep.slope<-function(chr, pos, model.list){
	mod<-model.list[[which(chr_names2==chr)]]
	sl<-nderiv(mod, pos)
	if(is.na(sl)){sl<-nderiv(mod, pos+1);if(is.na(sl)){sl<-nderiv(mod, pos-1)}}
	return(sl)
}

# determine slope at each pos
# rep.dynamics.hap$rep.slope<-sapply(1:nrow(rep.dynamics.hap),function(r){
# 	find.rep.slope(rep.dynamics.hap$chr[r],rep.dynamics.hap$pos[r],loess.models.hap)
# })
# rep.dynamics.dip$rep.slope<-sapply(1:nrow(rep.dynamics.dip),function(r){
# 	find.rep.slope(rep.dynamics.dip$chr[r],rep.dynamics.dip$pos[r],loess.models.dip)
# })

rep.dynamics$rep.slope<-sapply(1:nrow(rep.dynamics),function(r){
	find.rep.slope(rep.dynamics$chr[r],rep.dynamics.hap$pos[r],loess.models)
})
```


```{r}
# ggplot() +
#   geom_line(aes(x=rep.dynamics.hap$pos[rep.dynamics.hap$chr_num==1], y=rep.dynamics.hap$ratio[rep.dynamics.hap$chr_num==1]), color='red') +
#    geom_line(aes(x=rep.dynamics.hap$pos[rep.dynamics.hap$chr_num==1], y=predict(loess.models.hap[[1]])), color='black') 
# 
# ggplot() +
#   geom_line(aes(x=rep.dynamics.dip$pos[rep.dynamics.dip$chr_num==1], y=rep.dynamics.dip$ratio[rep.dynamics.dip$chr_num==1]), color='blue') +
#  
#   geom_line(aes(x=rep.dynamics.dip$pos[rep.dynamics.dip$chr_num==1], y=predict(loess.models.dip[[1]])), color='black')


# df %>% filter(chr_num == 1) %>% ggplot() +
#   geom_line(aes(x=pos, y=rep.slope.hap), color='red') +
#   geom_line(aes(x=pos, y=rep.slope.dip), color='blue')
```


```{r}
df <- df %>% select(-rep.ratio.dip, -rep.ratio.hap, -rep.strand.dip, -rep.strand.hap, -rep.slope.hap, -rep.slope.dip)
```

```{r}
# df <- df %>% left_join(rep.dynamics.hap[,c("pos","chr_num", "ratio", "rep.slope")], by=c("chr_num", "pos500"="pos"))
# 
# df <- df %>% rename("rep.ratio.hap" = "ratio",
#                             "rep.slope.hap" = "rep.slope")
# 
# df <- df %>% left_join(rep.dynamics.dip[,c("pos","chr_num", "ratio", "rep.slope")], by=c("chr_num", "pos500"="pos"))
# 
# df <- df %>% rename("rep.ratio.dip" = "ratio",
#                             "rep.slope.dip" = "rep.slope")

df <- df %>% left_join(rep.dynamics[,c("pos","chr_num", "ratio", "rep.slope")], by=c("chr_num", "pos500"="pos"))

# df$rep.strand.hap = sign(df$rep.slope.hap)
# df$rep.strand.dip = sign(df$rep.slope.dip)
df$rep.strand = sign(df$rep.slope)
```

```{r}
# df %>% filter(chr_num == 1) %>% ggplot() +
#   geom_line(aes(x=pos, y=rep.strand.hap), color='red') +
#   geom_line(aes(x=pos, y=rep.strand.dip), color='blue')

df %>% filter(chr_num == 1) %>% ggplot() +
  geom_line(aes(x=pos, y=rep.strand), color='red') 

```


Thinking about collisions again

```{r}
collisions <- df %>% select(chr_num, pos, rep.strand, net_tot.minus, net_tot.plus, mutated)
```

So for the transcription direction, I am using the direction of whatever is greater between net_tot.plus and net_tot.minus. Finding all the collisions spots.

```{r}
collisions <- collisions %>% mutate(tr.strand = case_when(
  net_tot.minus > net_tot.plus ~ -1,
  net_tot.plus > net_tot.minus ~ 1
))
```


Headon collisions: sign of replication and transcription is same
Cotranscriptional collisions: sign of replication and transcription is diff

```{r}
collisions <- collisions %>% mutate(collision = case_when(
  rep.strand * tr.strand == 1 ~ 1,
  TRUE ~ 0
))

collisions <- collisions %>% mutate(net_tot = case_when(
  tr.strand == 1 ~ net_tot.plus,
  tr.strand == -1 ~ net_tot.minus,
  TRUE ~ 0
))

collisions$class <- df$class
collisions <- collisions %>% mutate(net_tot.scaled = scale(net_tot))
collisions <- collisions %>% mutate(collision.scaled = scale(collision))
# collisions <- collisions %>% mutate(net_tot.scaled = scale(net_tot))
```

```{r}
# not significant
mod <- glm(mutated ~ net_tot.scaled + collision.scaled, family = "binomial", data=collisions)
summary(mod)
```

```{r}
mod <- glm(mutated ~ collision.scaled + net_tot.scaled + class, family = "binomial", collisions)
summary(mod)

# Running a wald test too!

wald.test(b=coef(mod), Sigma = vcov(mod), Terms=4:20)
```

```{r}
mod <- glm(mutated ~ collision.scaled + class, family = "binomial", collisions)
summary(mod)
```

```{r}
collisions$ref <- df$ref
collisions$rep.ratio <- df$ratio

collisions$rep.ratio <- replace_na(collisions$rep.ratio, 1)
collisions$rep.ratio.scaled <- scale(collisions$rep.ratio)
```

```{r}
mod <- glm(mutated ~ collision.scaled + net_tot.scaled + ref + rep.ratio.scaled, family = "binomial", collisions)
summary(mod)
```

```{r}
collisions <- collisions %>% mutate(ORF = ifelse(class == "ORF" | class == "ORFs", 1, 0))
collisions <- collisions %>% mutate(CDS = ifelse(class == "CDS", 1, 0))

collisions$ORF.scaled = scale(collisions$ORF)
collisions$CDS.scaled = scale(collisions$CDS)

collisions$type <- df$TYPE

collisions <- collisions %>% mutate(SNM = ifelse(type == "SNP", 1, 0))
collisions$SNM <- replace_na(collisions$SNM,0)
collisions <- collisions %>% mutate(INDEL = ifelse(type == "INDEL", 1, 0))
collisions$INDEL <- replace_na(collisions$INDEL, 0)

collisions <- collisions %>% mutate(ref.type = case_when(
  ref == "A" | ref == "T" ~ "A/T",
  ref == "G" | ref == "C" ~ "G/C",
  TRUE ~ NA
))
```

```{r}
mod <- glm(SNM ~ collision.scaled + net_tot.scaled + ref.type + rep.ratio.scaled + CDS.scaled, family = "binomial", collisions)
summary(mod)
```

```{r}
mod <- glm(INDEL ~ collision.scaled + net_tot.scaled + ref.type + rep.ratio.scaled + CDS.scaled, family = "binomial", collisions)
summary(mod)
```

```{r}
mod <- glm(SNM ~ collision.scaled * net_tot.scaled + ref.type + rep.ratio.scaled + CDS.scaled, family = "binomial", collisions)
summary(mod)
```

```{r}
mod <- glm(SNM ~ collision.scaled + net_tot.scaled * CDS.scaled + ref.type + rep.ratio.scaled , family = "binomial", collisions)
summary(mod)
```

Call:
glm(formula = SNM ~ collision.scaled + net_tot.scaled * CDS.scaled + 
    ref.type + rep.ratio.scaled, family = "binomial", data = collisions)

Coefficients:
                           Estimate Std. Error  z value Pr(>|z|)    
(Intercept)               -9.272479   0.037788 -245.382   <2e-16 ***
collision.scaled           0.015059   0.022914    0.657   0.5111    
net_tot.scaled             0.020352   0.015757    1.292   0.1965    
CDS.scaled                -0.002899   0.024021   -0.121   0.9039    
ref.typeG/C                1.011537   0.047563   21.267   <2e-16 ***
rep.ratio.scaled          -0.032970   0.023185   -1.422   0.1550    
net_tot.scaled:CDS.scaled  0.032636   0.015224    2.144   0.0321 *  
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 37058  on 12071325  degrees of freedom
Residual deviance: 36574  on 12071319  degrees of freedom
AIC: 36588

Number of Fisher Scoring iterations: 12

```{r}
mod <- glm(SNM ~ net_tot.scaled * CDS.scaled + ref.type + rep.ratio.scaled , family = "binomial", collisions)
summary(mod)
```

Call:
glm(formula = SNM ~ net_tot.scaled * CDS.scaled + ref.type + 
    rep.ratio.scaled, family = "binomial", data = collisions)

Coefficients:
                           Estimate Std. Error  z value Pr(>|z|)    
(Intercept)               -9.272484   0.037787 -245.386   <2e-16 ***
net_tot.scaled             0.022122   0.014665    1.508   0.1314    
CDS.scaled                -0.001877   0.023964   -0.078   0.9376    
ref.typeG/C                1.011931   0.047559   21.277   <2e-16 ***
rep.ratio.scaled          -0.032264   0.023151   -1.394   0.1634    
net_tot.scaled:CDS.scaled  0.032376   0.015200    2.130   0.0332 *  
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 37058  on 12071325  degrees of freedom
Residual deviance: 36575  on 12071320  degrees of freedom
AIC: 36587

Number of Fisher Scoring iterations: 12

Got some significant stuff in here

```{r}
mod <- glm(SNM ~ net_tot.scaled * CDS.scaled + ref.type , family = "binomial", collisions)
summary(mod)
```

Call:
glm(formula = SNM ~ net_tot.scaled * CDS.scaled + ref.type, family = "binomial", 
    data = collisions)

Coefficients:
                           Estimate Std. Error  z value Pr(>|z|)    
(Intercept)               -9.272137   0.037782 -245.409   <2e-16 ***
net_tot.scaled             0.021689   0.015006    1.445   0.1484    
CDS.scaled                -0.002244   0.023973   -0.094   0.9254    
ref.typeG/C                1.012341   0.047558   21.286   <2e-16 ***
net_tot.scaled:CDS.scaled  0.032397   0.015225    2.128   0.0334 *  
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 37058  on 12071325  degrees of freedom
Residual deviance: 36577  on 12071321  degrees of freedom
AIC: 36587

Number of Fisher Scoring iterations: 12

Trying for indels

```{r}
mod <- glm(INDEL ~ collision.scaled + ref.type + rep.ratio.scaled , family = "binomial", collisions)
summary(mod)
```

Call:
glm(formula = INDEL ~ collision.scaled + ref.type + rep.ratio.scaled, 
    family = "binomial", data = collisions)

Coefficients:
                  Estimate Std. Error z value Pr(>|z|)    
(Intercept)      -11.78774    0.13300 -88.627   <2e-16 ***
collision.scaled  -0.06179    0.10531  -0.587    0.557    
ref.typeG/C        0.19804    0.20201   0.980    0.327    
rep.ratio.scaled   0.09514    0.09828   0.968    0.333    
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 2540.2  on 12071325  degrees of freedom
Residual deviance: 2538.1  on 12071322  degrees of freedom
AIC: 2546.1

Number of Fisher Scoring iterations: 14

Trying to make a large model and working my way backwards. Making another testing dataframe

```{r}
test.df <- collisions[,c("chr_num", "pos", "rep.strand", "mutated", "collision", "net_tot", "rep.ratio", "CDS", "SNM", "INDEL", "ref.type")]
test.df$pwr.tot <- df$pwr.tot
test.df$rep.slope <- df$rep.slope
```

Making a scaled dataset

```{r}
df.scaled <- test.df[, c("chr_num", "pos", "mutated", "SNM", "INDEL", "ref.type", "CDS", "rep.strand")]
# df.scaled$rep.strand <- scale(test.df$rep.strand)
df.scaled$collision <- scale(test.df$collision)
df.scaled$net_tot <- scale(test.df$net_tot)
df.scaled$rep.ratio <- scale(test.df$rep.ratio)
df.scaled$rep.slope <- scale(test.df$rep.slope)
df.scaled$pwr.tot <- scale(test.df$pwr.tot)
# df.scaled$CDS <- scale(test.df$CDS)
```

Model making (yay!)

```{r}
mod <- glm(SNM ~ CDS + ref.type + rep.strand + collision + net_tot + rep.ratio + rep.slope + pwr.tot, family = "binomial", df.scaled)
summary(mod)
```

Call:
glm(formula = SNM ~ CDS + ref.type + rep.strand + collision + 
    net_tot + rep.ratio + rep.slope + pwr.tot, family = "binomial", 
    data = df.scaled)

Coefficients:
             Estimate Std. Error  z value Pr(>|z|)    
(Intercept) -9.317795   0.041932 -222.215  < 2e-16 ***
CDS          0.079793   0.303209    0.263  0.79243    
ref.typeG/C  1.006715   0.048016   20.966  < 2e-16 ***
rep.strand   0.057382   0.040279    1.425  0.15427    
collision   -0.008878   0.022966   -0.387  0.69909    
net_tot      0.013820   0.016982    0.814  0.41575    
rep.ratio   -0.070053   0.025625   -2.734  0.00626 ** 
rep.slope    0.012607   0.041170    0.306  0.75943    
pwr.tot      0.434780   0.072157    6.025 1.69e-09 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 36219  on 11364703  degrees of freedom
Residual deviance: 35668  on 11364695  degrees of freedom
  (706622 observations deleted due to missingness)
AIC: 35686

Number of Fisher Scoring iterations: 12

```{r}
mod <- glm(SNM ~ CDS + ref.type + rep.strand + collision + net_tot + rep.ratio + rep.slope, family = "binomial", df.scaled)
summary(mod)
```

Call:
glm(formula = SNM ~ CDS + ref.type + rep.strand + collision + 
    net_tot + rep.ratio + rep.slope, family = "binomial", data = df.scaled)

Coefficients:
             Estimate Std. Error  z value Pr(>|z|)    
(Intercept) -9.231493   0.038225 -241.506  < 2e-16 ***
CDS          0.091979   0.303229    0.303  0.76164    
ref.typeG/C  1.014777   0.048009   21.137  < 2e-16 ***
rep.strand   0.059566   0.040226    1.481  0.13866    
collision   -0.003786   0.022960   -0.165  0.86904    
net_tot      0.014484   0.016703    0.867  0.38586    
rep.ratio   -0.073347   0.025514   -2.875  0.00404 ** 
rep.slope    0.012076   0.041179    0.293  0.76932    
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 36219  on 11364703  degrees of freedom
Residual deviance: 35725  on 11364696  degrees of freedom
  (706622 observations deleted due to missingness)
AIC: 35741

Number of Fisher Scoring iterations: 12

```{r}
mod <- glm(SNM ~ ref.type + rep.strand + collision * net_tot * CDS + rep.ratio + rep.slope, family = "binomial", df.scaled)
summary(mod)
```

Call:
glm(formula = SNM ~ ref.type + rep.strand + collision * net_tot * 
    CDS + rep.ratio + rep.slope, family = "binomial", data = df.scaled)

Coefficients:
                       Estimate Std. Error  z value Pr(>|z|)    
(Intercept)           -9.237201   0.038624 -239.155  < 2e-16 ***
ref.typeG/C            1.014837   0.048010   21.138  < 2e-16 ***
rep.strand             0.058551   0.040244    1.455  0.14570    
collision             -0.006513   0.024812   -0.263  0.79293    
net_tot                0.016362   0.018464    0.886  0.37552    
CDS                   -0.062120   0.049053   -1.266  0.20538    
rep.ratio             -0.073939   0.025553   -2.894  0.00381 ** 
rep.slope              0.011144   0.041209    0.270  0.78683    
collision:net_tot      0.005071   0.017718    0.286  0.77470    
collision:CDS          0.036917   0.024511    1.506  0.13203    
net_tot:CDS            0.047555   0.019888    2.391  0.01680 *  
collision:net_tot:CDS -0.010370   0.012599   -0.823  0.41046    
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 36219  on 11364703  degrees of freedom
Residual deviance: 35719  on 11364692  degrees of freedom
  (706622 observations deleted due to missingness)
AIC: 35743

Number of Fisher Scoring iterations: 12

```{r}
mod <- glm(SNM ~ ref.type + rep.strand + collision + net_tot * CDS + rep.ratio + rep.slope, family = "binomial", df.scaled)
summary(mod)
```

Call:
glm(formula = SNM ~ ref.type + rep.strand + collision + net_tot * 
    CDS + rep.ratio + rep.slope, family = "binomial", data = df.scaled)

Coefficients:
            Estimate Std. Error  z value Pr(>|z|)    
(Intercept) -9.23216    0.03820 -241.679  < 2e-16 ***
ref.typeG/C  1.01469    0.04801   21.135  < 2e-16 ***
rep.strand   0.05940    0.04023    1.476  0.13982    
collision   -0.00284    0.02301   -0.123  0.90175    
net_tot      0.01487    0.01763    0.844  0.39889    
CDS         -0.01135    0.02528   -0.449  0.65352    
rep.ratio   -0.07352    0.02551   -2.882  0.00395 ** 
rep.slope    0.01186    0.04118    0.288  0.77338    
net_tot:CDS  0.03383    0.01550    2.183  0.02903 *  
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 36219  on 11364703  degrees of freedom
Residual deviance: 35722  on 11364695  degrees of freedom
  (706622 observations deleted due to missingness)
AIC: 35740

Number of Fisher Scoring iterations: 12

```{r}
mod <- glm(SNM ~ ref.type + collision + net_tot * CDS + rep.ratio + rep.strand, family = "binomial", df.scaled)
summary(mod)
```

Call:
glm(formula = SNM ~ ref.type + collision + net_tot * CDS + rep.ratio + 
    rep.strand, family = "binomial", data = df.scaled)

Coefficients:
             Estimate Std. Error  z value Pr(>|z|)    
(Intercept) -9.231173   0.038218 -241.537  < 2e-16 ***
ref.typeG/C  1.014708   0.048009   21.136  < 2e-16 *** (more mutations at GC sites)
collision   -0.002812   0.023006   -0.122  0.90271    
net_tot      0.012387   0.017667    0.701  0.48322    
CDS         -0.161784   0.358608   -0.451  0.65189    
rep.ratio   -0.074869   0.025082   -2.985  0.00284 ** (more ratio, early replication, less mutation -  higher mutation rates in late replicating region - idk why)
rep.strand   0.068667   0.024161    2.842  0.00448 ** (idk)
net_tot:CDS  0.480126   0.219810    2.184  0.02894 *  (interaction between transcription and CDS. if transcription high, high mutation in coding sequences)

subset: split into CDS = true and CDS = false and run model separately
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 36219  on 11364703  degrees of freedom
Residual deviance: 35722  on 11364696  degrees of freedom
  (706622 observations deleted due to missingness)
AIC: 35738

Number of Fisher Scoring iterations: 12

```{r}
df.CDS <- df.scaled %>% filter(CDS == 1)
df.noCDS <- df.scaled %>% filter(CDS == 0)
```


```{r}
mod <- glm(SNM ~ ref.type + collision + net_tot + rep.ratio + rep.strand, family = "binomial", df.CDS)
summary(mod)
```

Call:
glm(formula = SNM ~ ref.type + collision + net_tot + rep.ratio + 
    rep.strand, family = "binomial", data = df.CDS)

Coefficients:
             Estimate Std. Error z value Pr(>|z|)    
(Intercept) -10.03733    0.73444 -13.667   <2e-16 ***
ref.typeG/C   1.31963    0.67720   1.949   0.0513 .  
collision     0.19405    0.29706   0.653   0.5136    
net_tot       0.47429    0.23492   2.019   0.0435 *  
rep.ratio     0.05759    0.32984   0.175   0.8614    
rep.strand    0.50949    0.45119   1.129   0.2588    
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 211.15  on 59609  degrees of freedom
Residual deviance: 198.89  on 59604  degrees of freedom
  (697 observations deleted due to missingness)
AIC: 210.89

Number of Fisher Scoring iterations: 12

```{r}
mod <- glm(SNM ~ ref.type + collision + net_tot + rep.ratio + rep.strand, family = "binomial", df.noCDS)
summary(mod)
```

Call:
glm(formula = SNM ~ ref.type + collision + net_tot + rep.ratio + 
    rep.strand, family = "binomial", data = df.noCDS)

Coefficients:
            Estimate Std. Error  z value Pr(>|z|)    
(Intercept) -9.22987    0.03826 -241.219  < 2e-16 ***
ref.typeG/C  1.01314    0.04814   21.048  < 2e-16 ***
collision   -0.00618    0.02317   -0.267  0.78967    
net_tot      0.01291    0.01743    0.741  0.45891    
rep.ratio   -0.07632    0.02516   -3.033  0.00242 ** 
rep.strand   0.06566    0.02423    2.710  0.00673 ** 
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 36008  on 11305093  degrees of freedom
Residual deviance: 35519  on 11305088  degrees of freedom
  (705925 observations deleted due to missingness)
AIC: 35531

Number of Fisher Scoring iterations: 12


Pretty plots time

```{r}
test.cds <- test.df %>% filter(CDS == 1)
test.nocds <- test.df %>% filter(CDS == 0)

Stats.cds <- test.cds %>% group_by(SNM) %>% summarize(N = n(),
  Mean = mean(net_tot), SD = sd(net_tot),
                                            CI_L = Mean - (SD * 1.96)/sqrt(N),
                                            CI_U = Mean + (SD * 1.96)/sqrt(N))

Stats.nocds <- test.nocds %>% group_by(SNM) %>% summarize(N = n(),
  Mean = mean(net_tot), SD = sd(net_tot),
                                            CI_L = Mean - (SD * 1.96)/sqrt(N),
                                            CI_U = Mean + (SD * 1.96)/sqrt(N))

test.cds %>% filter(net_tot != 0) %>% ggplot() +
  geom_violin(aes(x=as.character(SNM), y=net_tot, fill = as.character(SNM)), show.legend = FALSE) +
  geom_point(mapping = aes(as.character(SNM), Mean), data = Stats.cds) +
  geom_errorbar(mapping = aes(x = as.character(SNM), ymin = CI_L, ymax = CI_U), 
                data = Stats.cds, width = 0.2) +
  scale_x_discrete(name ="SNM in Coding Sequence", 
                    labels=c("Absent", "Present")) +
  ylab("Total Number of Transcripts") +
  scale_y_log10(limits = c(1, max(test.df$net_tot))) 
  

test.nocds %>% filter(net_tot != 0) %>% ggplot() +
  geom_violin(aes(x=as.character(SNM), y=net_tot, fill = as.character(SNM)), show.legend = FALSE) +
  geom_point(mapping = aes(as.character(SNM), Mean), data = Stats.nocds) +
  geom_errorbar(mapping = aes(x = as.character(SNM), ymin = CI_L, ymax = CI_U), 
                data = Stats.nocds, width = 0.2) +
  scale_x_discrete(name ="SNM in Non-Coding Sequence", 
                    labels=c("Absent", "Present")) +
  ylab("Total Number of Transcripts") +
  scale_y_log10(limits = c(1, max(test.df$net_tot)))

```

```{r}

Stats.ratio <- test.nocds %>% group_by(SNM) %>% summarize(N = n(),
  Mean = mean(rep.ratio), SD = sd(rep.ratio),
                                            CI_L = Mean - (SD * 1.96)/sqrt(N),
                                            CI_U = Mean + (SD * 1.96)/sqrt(N))


ggplot() + 
  geom_violin(aes(x=as.character(SNM), y=rep.ratio, fill = as.character(SNM)), data = test.df, show.legend = FALSE) +
  geom_point(mapping = aes(as.character(SNM), Mean), data = Stats.ratio) +
  geom_errorbar(mapping = aes(x = as.character(SNM), ymin = CI_L, ymax = CI_U), 
                data = Stats.ratio, width = 0.2) +
  scale_x_discrete(name ="SNM", 
                    labels=c("Absent", "Present")) +
  ylab("Replication Ratio")
  
```


```{r}
ggplot() +
  geom_density(aes(x=rep.ratio, color="SNM"), alpha=0.5, data = test.df %>% filter(SNM == 1)) +
 geom_density(aes(x=rep.ratio, color="NonSNM"), alpha=0.5, data = test.df %>% filter(SNM == 0)) +
  xlim(c(1,2)) +
  xlab("Replication Ratio") +
  ylab("Density") +
  scale_color_manual("", 
                      values = c("SNM"="red", 
                                 "NonSNM"="blue"))
```

```{r}
ggplot() +
  geom_histogram(aes(x=rep.ratio), fill='red', alpha=0.5, data = test.df %>% filter(SNM == 1))
```


Models, models and more models:

```{r}
mod <- glm(SNM ~ ref.type, family = "binomial", df.scaled)
summary(mod) 
```

Call:
glm(formula = SNM ~ ref.type, family = "binomial", data = df.scaled)

Coefficients:
            Estimate Std. Error z value Pr(>|z|)    
(Intercept) -9.27090    0.03777  -245.4   <2e-16 ***
ref.typeG/C  1.01301    0.04756    21.3   <2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 37058  on 12071325  degrees of freedom
Residual deviance: 36582  on 12071324  degrees of freedom
AIC: 36586

Number of Fisher Scoring iterations: 12

```{r}
df.scaled %>% group_by(ref.type, SNM) %>%
  summarize(n = n()) %>% filter(SNM == 1)
```



Trying the raw netseq data

Key for chromosome names

```{r}
chr_names<-c("chrI","chrII","chrIII","chrIV","chrV","chrVI","chrVII","chrVIII", "chrIX","chrX","chrXI","chrXII","chrXIII","chrXIV","chrXV","chrXVI")

chr_names2<-c("chr01","chr02","chr03","chr04","chr05","chr06","chr07","chr08","chr09","chr10","chr11","chr12","chr13","chr14","chr15","chr16")
```

Importing the transcription data. Source is the Churchman and Weissman paper.  

```{r}
net.plus<-read.delim("transcription_data/GSM617027_WT_NC_plus.txt",h=F,sep=" ")
names(net.plus)<-c("pos","value")
net.minus<-read.delim("transcription_data/GSM617027_WT_NC_minus.txt",h=F,sep=" ")
names(net.minus)<-c("pos","value")
```

Chromosome annotation
1. Make a vector of the same length as the data, add 0 as the first element and chuck the last element.  
2. For each chromosome, the files are organized in ascending order of pos. So, difference between the pos and prev series will be positive everywhere except where there is a chromosome change (starting from a small value again)  
3. Those are the chromosome breaks  
4. Repeat for the other dataset

```{r}
# need to add chromosome annotation (chromosomes are in order)
prev<-c(0,net.plus$pos[1:(nrow(net.plus)-1)])
diff<-sign(net.plus$pos-prev)
net.plus$chr_num<-NA
cut.rows<-c(1,which(diff<0),nrow(net.plus))
for(i in c(1:16)){net.plus$chr_num[cut.rows[i]:cut.rows[i+1]]<-i}


prev<-c(0,net.minus$pos[1:(nrow(net.minus)-1)])
diff<-sign(net.minus$pos-prev)
net.minus$chr_num<-NA
cut.rows<-c(1,which(diff<0),nrow(net.minus))
for(i in c(1:16)){net.minus$chr_num[cut.rows[i]:cut.rows[i+1]]<-i}
```

Rescaling the data to be integers. The data seems to have been multiplied by some constant to convert the number of reads data to something else. I am not sure where the multiplication constant is coming from but it is the minimum value in the dataset.

```{r}
val.plus <- min(net.plus$value)
val.minus <- min(net.minus$value)

# re-scale y so that it is an integer (assumes lowest non-zero value of y corresponds to 1 read per 10^7 sequences)
net.plus$reads<-net.plus$value/val.plus
net.minus$reads<-net.minus$value/val.minus

```

Exploratory plots. Plot of reads across chromosome I and a zoomed-in version of what one of the genes looks like

```{r}
net.plus %>% filter(chr_num == 1) %>% ggplot() + geom_line(aes(x=pos, y=reads)) +
  scale_y_log10()
net.plus %>% filter(chr_num == 1) %>% filter(pos > 70000 & pos < 75000) %>% ggplot() + geom_line(aes(x=pos, y=reads)) +
  scale_y_log10()
```


```{r}
net <- full_join(net.plus, net.minus, by = c('chr_num', 'pos'), suffix = c('.plus', '.minus'))
net$reads.minus <- replace_na(net$reads.minus, 0)
net$reads.plus <- replace_na(net$reads.plus, 0)
net$tot <- net$reads.minus + net$reads.plus
net <- net %>% select(chr_num, pos, tot)

df.scaled <- df.scaled %>% left_join(net, by = c('chr_num', 'pos'))
df.scaled$tot <- replace_na(df.scaled$tot, 0)
df.scaled$tot <- scale(df.scaled$tot)
```

```{r}
mod <- glm(SNM ~ tot*CDS + rep.ratio + collision + rep.strand, family = "binomial", df.scaled)
summary(mod)
```


```{r}
mod <- glm(SNM ~ ref.type + collision + net_tot * CDS + rep.ratio + rep.strand, family = "binomial", df.scaled)
summary(mod)
```








```{r}
mod <- glm(INDEL ~ ref.type + collision + net_tot * CDS + rep.ratio + rep.strand, family = "binomial", df.scaled)
summary(mod)
```

Call:
glm(formula = INDEL ~ ref.type + collision + net_tot * CDS + 
    rep.ratio + rep.strand, family = "binomial", data = df.scaled)

Coefficients:
              Estimate Std. Error z value Pr(>|z|)    
(Intercept) -11.788588   1.704816  -6.915 4.68e-12 ***
ref.typeG/C   0.177514   0.203386   0.873    0.383    
collision    -0.078418   0.108047  -0.726    0.468    
net_tot       0.007416   1.944418   0.004    0.997    
CDS          -0.836096  23.985612  -0.035    0.972    
rep.ratio     0.064900   0.106013   0.612    0.540    
rep.strand    0.109750   0.105029   1.045    0.296    
net_tot:CDS  -0.001497  27.412205   0.000    1.000    
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 2504.9  on 11364703  degrees of freedom
Residual deviance: 2501.3  on 11364696  degrees of freedom
  (706622 observations deleted due to missingness)
AIC: 2517.3

Number of Fisher Scoring iterations: 22
