---
title: "Analysis Closest"
author: "Rutuja"
date: "2024-04-01"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Import statements

```{r}
library(tidyverse)
```

Required keys

```{r}
chr_names<-c("chrI","chrII","chrIII","chrIV","chrV","chrVI","chrVII","chrVIII", "chrIX","chrX","chrXI","chrXII","chrXIII","chrXIV","chrXV","chrXVI")

chr_names2<-c("chr01","chr02","chr03","chr04","chr05","chr06","chr07","chr08","chr09","chr10","chr11","chr12","chr13","chr14","chr15","chr16")
```

# Reading in files

```{r}
# power to detect mutations
power<-read.delim("power_by_site_with_bases.txt",h=F)
names(power)<-c("chr","pos","hap.wt","hap.del","dip.wt","dip.del","ref")
power$chr_num<-sapply(power$chr,function(i){which(chr_names2==i)})
d <- power %>% select(chr_num, pos, ref, hap.wt, hap.del, dip.wt, dip.del)
d$pwr.tot<-rowSums(d[,c(4:7)])
```

```{r}
net <- read.csv("net_mapped/closest_net.csv")
```

```{r}
df.plus <- net[net$strand == "+",]
df.minus <- net[net$strand == "-",]

d <- d %>% left_join(df.minus[,c("pos", "chr_num", "total", "feature")], by=c('pos', 'chr_num'), suffix = c("", ".minus"))
d <- d %>% rename("net_tot.minus" = "total",
                            "class.minus" = "feature")
d <- d %>% left_join(df.plus[,c("pos", "chr_num", "total", "feature")], by=c('pos', 'chr_num'), suffix = c("", ".plus"))
d <- d %>% rename("net_tot.plus" = "total",
                            "class.plus" = "feature")

d$pos500 <- ((d$pos-1) %/% 1000) * 1000 + 501
```


Starting with the replication dynamics data first

```{r}
rep.dynamics <- read.csv("replication_data/clean_rep.csv")
```

```{r}
d <- d %>% left_join(rep.dynamics[,c("pos","chr_num", "ratio", "rep.slope", "rep.sign")], by=c("chr_num", "pos500"="pos"))
```

# Now combining mutation data

Reading in the mutation data

```{r}
mut <- load("mutation_data/Sharp_etal_variants")
```

Filter to keep specific columns

```{r}
test.rv <- rv %>% select(CHROM, POS, REF, ALT, TYPE, RDH, MAT, ploidy, allele.freq, chr_copy_num, aneu11, indel.type, genic, mut.type, NCALLED)

test.rv$chr_num <- sapply(test.rv$CHROM,function(i){which(chr_names2==i)})
test.rv <- test.rv %>% rename("pos" = "POS")
```

Final join

```{r}
d <- d %>% left_join(test.rv, by = c("chr_num", "pos"))
```

# Generating more useful calculated columns

```{r}
d$net_tot.minus <- d$net_tot.minus %>% replace_na(0)
d$net_tot.plus <- d$net_tot.plus %>% replace_na(0)
d$net_tot <- d$net_tot.minus + d$net_tot.plus
d <- d %>% mutate(tr.or.not = case_when(
  !is.na(class.minus) | !is.na(class.plus) ~ TRUE,
  TRUE ~ FALSE
))

d <- d %>% mutate(class = case_when(
  !is.na(class.minus) & is.na(class.plus) ~ class.minus,
  is.na(class.minus) & !is.na(class.plus) ~ class.plus,
  !is.na(class.minus) & !is.na(class.plus) ~ "both",
  TRUE ~ "none"
))

d <- d %>% mutate(transcribed = ifelse(d$net_tot > 0, TRUE, FALSE))

df <- d
rm(d)

df <- df %>% select(-CHROM)
df$net_tot.minus <- replace_na(df$net_tot.minus, 0)
df$net_tot.plus <- replace_na(df$net_tot.plus, 0)
df$class.minus <- replace_na(df$class.minus, "none")
df$class.plus <- replace_na(df$class.plus, "none")
df$net_tot.minus <- replace_na(df$net_tot.minus, 0)
df$net_tot.plus <- replace_na(df$net_tot.plus, 0)
df$allele.freq <- replace_na(df$allele.freq, 0)
df$genic <- replace_na(df$genic, FALSE)
df$NCALLED <- replace_na(df$NCALLED, 0)
df <- df %>% mutate(mutated = ifelse(df$NCALLED > 0, 1, 0))
df$rep.strand <- df$rep.sign

ready.cols <- sapply(df, function(x)all(!is.na(x)))
ready.cols[ready.cols]
```

Adding some collision info here

```{r}
collisions <- df %>% select(chr_num, pos, rep.strand, net_tot.minus, net_tot.plus, mutated)
```

So for the transcription direction, I am using the direction of whatever is greater between net_tot.plus and net_tot.minus. Finding all the collisions spots.

```{r}
collisions <- collisions %>% mutate(tr.strand = case_when(
  net_tot.minus > net_tot.plus ~ -1,
  net_tot.plus > net_tot.minus ~ 1,
  TRUE ~ 0
))

collisions <- collisions %>% mutate(collision = case_when(
  rep.strand * tr.strand == 1 ~ 1,
  TRUE ~ 0
))

collisions <- collisions %>% mutate(net_tot = net_tot.minus + net_tot.minus)

collisions$class <- df$class
collisions <- collisions %>% mutate(ORF = ifelse(class == "ORF" | class == "ORFs", 1, 0))
collisions <- collisions %>% mutate(CDS = ifelse(class == "CDS", 1, 0))

collisions$ORF.scaled = scale(collisions$ORF)
collisions$CDS.scaled = scale(collisions$CDS)

collisions$type <- df$TYPE

collisions <- collisions %>% mutate(SNM = ifelse(type == "SNP", 1, 0))
collisions$SNM <- replace_na(collisions$SNM,0)
collisions <- collisions %>% mutate(INDEL = ifelse(type == "INDEL", 1, 0))
collisions$INDEL <- replace_na(collisions$INDEL, 0)

collisions$ref <- df$ref
collisions$rep.ratio <- df$ratio

collisions$rep.ratio <- replace_na(collisions$rep.ratio, 1)
collisions$rep.ratio.scaled <- scale(collisions$rep.ratio)
collisions <- collisions %>% mutate(ref.type = case_when(
  ref == "A" | ref == "T" ~ "A/T",
  ref == "G" | ref == "C" ~ "G/C",
  TRUE ~ NA
))
```

Trying to make a large model and working my way backwards. Making another testing dataframe

```{r}
test.df <- collisions[,c("chr_num", "pos", "rep.strand", "mutated", "collision", "net_tot", "rep.ratio", "CDS", "SNM", "INDEL", "ref.type")]
test.df$pwr.tot <- df$pwr.tot
test.df$rep.slope <- df$rep.slope
```

Making a scaled dataset

```{r}
df.scaled <- test.df[, c("chr_num", "pos", "mutated", "SNM", "INDEL", "ref.type", "CDS", "rep.strand", "collision", "net_tot", "rep.ratio", "CDS")]
# df.scaled$rep.strand <- scale(test.df$rep.strand)
# df.scaled$collision <- scale(test.df$collision)
# df.scaled$net_tot <- scale(test.df$net_tot)
# df.scaled$rep.ratio <- scale(test.df$rep.ratio)
df.scaled$rep.slope <- scale(test.df$rep.slope)
df.scaled$pwr.tot <- scale(test.df$pwr.tot)
# df.scaled$CDS <- scale(test.df$CDS)
```



```{r}
write.csv(df.scaled, "processed_data/closest.csv")
```

Model making (yay!)

```{r}
mod <- glm(SNM ~ ref.type + collision * net_tot * CDS + rep.ratio + rep.strand, family = "binomial", df.scaled)
summary(mod)
```

Call:
glm(formula = SNM ~ ref.type + collision * net_tot * CDS + rep.ratio + 
    rep.strand, family = "binomial", data = df.scaled)

Coefficients:
                        Estimate Std. Error z value Pr(>|z|)    
(Intercept)           -8.771e+00  1.215e-01 -72.162  < 2e-16 ***
ref.typeG/C            1.021e+00  4.782e-02  21.353  < 2e-16 ***
collision              3.193e-02  6.692e-02   0.477   0.6333    
net_tot                4.571e-06  6.585e-06   0.694   0.4876    
CDS                   -8.142e-01  7.369e-01  -1.105   0.2692    
rep.ratio             -3.362e-01  8.451e-02  -3.979 6.92e-05 ***
rep.strand            -1.589e-04  2.411e-02  -0.007   0.9947    
collision:net_tot     -2.093e-05  1.318e-05  -1.588   0.1122    
collision:CDS          1.391e+00  8.200e-01   1.697   0.0897 .  
net_tot:CDS            7.924e-05  4.124e-05   1.922   0.0546 .  
collision:net_tot:CDS -1.204e+00  1.672e+01  -0.072   0.9426    
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 36561  on 11449703  degrees of freedom
Residual deviance: 36054  on 11449693  degrees of freedom
  (621622 observations deleted due to missingness)
AIC: 36076

Number of Fisher Scoring iterations: 25

```{r}
mod <- glm(SNM ~ ref.type + net_tot + CDS + collision * rep.ratio + rep.strand, family = "binomial", df.scaled)
summary(mod)
```

Call:
glm(formula = SNM ~ ref.type + net_tot + CDS + collision * rep.ratio + 
    rep.strand, family = "binomial", data = df.scaled)

Coefficients:
                      Estimate Std. Error z value Pr(>|z|)    
(Intercept)         -8.796e+00  1.328e-01 -66.257  < 2e-16 ***
ref.typeG/C          1.021e+00  4.782e-02  21.358  < 2e-16 ***
net_tot             -2.989e-06  5.253e-06  -0.569 0.569394    
CDS                  1.377e-01  3.034e-01   0.454 0.649796    
collision            1.453e-01  3.069e-01   0.474 0.635746    
rep.ratio           -3.152e-01  9.307e-02  -3.387 0.000708 ***
rep.strand           1.419e-02  2.311e-02   0.614 0.539168    
collision:rep.ratio -1.087e-01  2.215e-01  -0.491 0.623415    
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 36561  on 11449703  degrees of freedom
Residual deviance: 36064  on 11449696  degrees of freedom
  (621622 observations deleted due to missingness)
AIC: 36080

Number of Fisher Scoring iterations: 12

```{r}
mod <- glm(SNM ~ net_tot + CDS + collision * rep.ratio + rep.strand, family = "binomial", df.scaled)
summary(mod)
```

```{r}
mod <- glm(SNM ~ ref.type + collision * net_tot + CDS + rep.ratio + rep.strand, family = "binomial", df.scaled)
summary(mod)
```

Call:
glm(formula = SNM ~ ref.type + collision + net_tot * CDS + rep.ratio + 
    rep.strand, family = "binomial", data = df.scaled)

Coefficients:
             Estimate Std. Error  z value Pr(>|z|)    
(Intercept) -9.231252   0.038074 -242.453  < 2e-16 ***
ref.typeG/C  1.021296   0.047819   21.358  < 2e-16 ***
collision    0.001137   0.022885    0.050    0.960    
net_tot     -0.018467   0.028639   -0.645    0.519    
CDS          0.006564   0.021427    0.306    0.759    
rep.ratio   -0.095309   0.024068   -3.960 7.49e-05 ***
rep.strand   0.021875   0.023141    0.945    0.344    
net_tot:CDS  0.020118   0.013231    1.521    0.128    
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 36561  on 11449703  degrees of freedom
Residual deviance: 36062  on 11449696  degrees of freedom
  (621622 observations deleted due to missingness)
AIC: 36078

Number of Fisher Scoring iterations: 12


